### ISSUE TITLE
Question about C3D_DORN training

### ISSUE THREAD

**Issue by youngskkim** [minghanz/cvo_ops#2]
Thank you for sharing your very impressive work.

I am trying to apply your work to train in the NuScenes dataset and found a few problems, and I am here to ask your intuition in training your code in another dataset.

First, the LiDAR configuration between KITTI and NS is quite different because of the number of the layer. (top: NS, bottom: KITTI)
![0_mask](https://user-images.githubusercontent.com/24770858/102422718-ba29ef00-404a-11eb-9b2e-72904572d4ad.png)
![0_mask](https://user-images.githubusercontent.com/24770858/102422744-c615b100-404a-11eb-9ed7-bd0fb646d168.png)
As shown in the figures, the number of points of NS is about 4000 and KITTI is 20000.
Moreover, the vertical resolution is especially different.
I am trying to tune neighbor_range and normal_nrange configuration but I couldn't figure out what does 2 and 9 mean yet..
So can you let me know what does 2 and 9 means? Does the range mean the euclidean distance?
Also, can you recommend how to change neighbor_range and normal_nrange configuration?

Second, I found the TODO in loss term in dorn.py code. (weight of loss_c3d: 1e-3)
Can you recommend the weight of c3d loss if there is some kind of rule of thumb?
For example, is it better to loss_dorn and loss_c3d has a similar loss? or loss_dorn should be bigger than loss_c3d?

Third, it seems that the c3d loss tends to fluctuate compared to dorn loss.
dorn loss converges from ~50 to 10 smoothly but c3d loss fluctuates between 100 ~ 10000.
I found it from both KITTI and NS, so I want to know if it is okay or if I did something wrong with my code.

Thank you for sharing your nice work again!

**Comment by minghanz** [comment]
@topys21c , you're welcome and thanks for your interest in this work. 

The parameters `neighbor_range` and `normal_nrange` are both in terms of pixels.  `neighbor_range` defines the maximum pixel offset when calculating the kernel function w.r.t. a certain point. In other words, the pairs of points with projected location on image too far away will not contribute to the continuous loss. 

`normal_nrange`  indicates the maximum pixel offset when calculating the normal vector of points in lidar point cloud. The normal direction at a point is calculated by the points in the neighborhood whose projected coordinate is within `normal_nrange` pixels of the reference point. I set `normal_nrange` to be larger than `neighbor_range` because reliable normal direction estimation on the sparse point cloud need more support points.

For sparser lidar point cloud, please consider set larger values for both parameters for the loss function to behave properly. 

I tend to set the weight of c3d loss so that it is in a similar magnitude with the original dorn loss. 

Yes, it is normal that the c3d loss fluctuates, because of the nature of the double sum and the fact that we are maximizing it I think. While the loss fluctuates, it still shows distinguishable effect on the final training result, so it is not behaving randomly. 

Please feel free to follow up if you have further questions.

**Comment by youngskkim** [comment]
@minghanz Thank you for the kind and fast answer!
Your comment helped me to understand the code better but I have further questions.

If I understand your paper correctly, the continuous loss (both hsv and xyz) should be calculated even when the projected sparse lidar points have a hole (glass of car).
**1.** Is neightbor_range of '2' is enough to contribute to the loss between pixels?
I found the pixel distance between points in KITTI is more than 3 pixels in the vertical direction.
In NuScnes, even I downsample by 2, pixel distance in vertical direction is more than 15. (3~5 in horizontal direction)
**2.** I am thinking to set normal_nrange around 20 because the normal vector should be calculated using both vertical and horizontal points,
and set neighbor_range around 5 because the HSV kernel should have a smaller size.
Do you think is it sounds reasonable?

In my opinion, the appropriate neighbor_range in my case should have a rectangle shape rather than a square considering the larger vertical gap between layers in NuScenes dataset.
I will try to modify the code to have a rectangle shape kernel if you agree with this idea.

I have one more question, 
**3.** is the res_mag_max parameter is for xyz geometric (exponential) kernel should be enlarged when the projected lidar points (gt) become more sparse than KITTI?

Sorry for too many questions at once.
I will look forward to your intuition!

**Comment by minghanz** [comment]
@topys21c No problem. 

1. `neightbor_range ` of 2 means a point is influenced by a 5*5 neighborhood around it, therefore for pixel distance about 3 pixels in KITTI is okay. Furthermore, the kernel of normal direction guides the nearby pixels to have consistent normal direction, therefore in reality the loss is effective for more than that range. 

2. Setting `normal_nrange` to 20 seems fine to me, if the point cloud is sparser than KITTI. To correct my above replying, larger `normal_nrange` may not necessarily increase the number of points when calculating normal direction. It is to guarantee that there are enough points covered. At most 8 points are used to calculate the normal direction. Therefore setting a large `normal_nrange` should not slow down the computation much. 

In terms of rectangular kernel, you can try it but I do not think it is necessary. It is okay that the vertical distance is larger. As long as you can cover more than one row of the scan in a neighborhood, you should be fine. Using a rectangular neighborhood may help reduce the number of points in the neighborhood. Other than that I do not see extra benefits. 
 
3. You do not need to change `res_mag_max` when the point cloud density changes.  It is used to regularize the behavior of kernel of normal directions.

**Comment by youngskkim** [comment]
@minghanz 
It seems the network is converging thanks to your comments.
I will share the results in NuScenes dataset and hyperparameters when the training is done.

~~I have another question in loss term.~~
~~In your dorn.py model, you subtract the c3d loss from dorn loss, because you want to minimize the dorn (cross entropy) and maximize the c3d (inner product) if I understand it correctly.~~

~~However, is it okay to use two losses that have contrary objectives? (one is to minimize and the other is maximize)~~
~~If dorn and c3d loss have similar value and if you sum them up, the total loss will be zero and even can be minus. Doesn't it affect to optimizer?~~

~~Isn't it better to formulate the c3d loss to be minimized? (for example, c3d loss = 1/inner_product so that the loss can be minimized and the inner product can be maximized)~~
~~I want to ask that near zero or minus loss does not affect the training.~~

I realized that it was a dumb question....
Instead of that dumb question, can I ask the final loss when you trained your model in KITTI?
Thank you!

**Comment by minghanz** [comment]
@topys21c What do you mean by final loss?
